runtime/compact.c

```diff
@@ -35,88 +35,59 @@
+   ...XXXyyyyyyyy00 becomes ...XXX01yyyyyy
+   Note that 01 corresponds to the "gray" color of the GC, so we can now
+   mix pointers and headers because there are no gray headers anywhere in
+   the heap (or outside) when we start a compaction (which must be done at
+   the end of a sweep phase).
 */
-#ifdef WITH_PROFINFO
-#define Make_ehd(s,t,c,p) \
-  (((s) << 10) | (t) << 2 | (c) | ((p) << PROFINFO_SHIFT))
-#else
-#define Make_ehd(s,t,c,p) (((s) << 10) | (t) << 2 | (c))
-#endif
-#define Whsize_ehd(h) Whsize_hd (h)
-#define Wosize_ehd(h) Wosize_hd (h)
-#define Tag_ehd(h) (((h) >> 2) & 0xFF)
-#ifdef WITH_PROFINFO
-#define Profinfo_ehd(hd) Profinfo_hd(hd)
-#endif
-#define Ecolor(w) ((w) & 3)
 
 typedef uintnat word;
 
+#define eptr(p) \
+  (((word) (p) & ~0x3FF) | ((((word) p) & 0x3FF) >> 2) | Caml_gray)
+#define dptr(p) ((word *) (((word) (p) & ~0x3FF) | ((((word) p) & 0xFF) << 2)))
+
 static void invert_pointer_at (word *p)
 {
   word q = *p;
-  CAMLassert (Ecolor ((intnat) p) == 0);
-
-  /* Use Ecolor (q) == 0 instead of Is_block (q) because q could be an
-     inverted pointer for an infix header (with Ecolor == 2). */
-  if (Ecolor (q) == 0 && Is_in_heap (q)){
-    switch (Ecolor (Hd_val (q))){
-    case 0:
-    case 3: /* Pointer or header: insert in inverted list. */
-      *p = Hd_val (q);
-      Hd_val (q) = (header_t) p;
-      break;
-    case 1: /* Infix header: make inverted infix list. */
-      /* Double inversion: the last of the inverted infix list points to
-         the next infix header in this block.  The last of the last list
-         contains the original block header. */
-      {
-        /* This block as a value. */
-        value val = (value) q - Infix_offset_val (q);
-        /* Get the block header. */
-        word *hp = (word *) Hp_val (val);
-
-        while (Ecolor (*hp) == 0) hp = (word *) *hp;
-        CAMLassert (Ecolor (*hp) == 3);
-        if (Tag_ehd (*hp) == Closure_tag){
-          /* This is the first infix found in this block. */
-          /* Save original header. */
-          *p = *hp;
-          /* Link inverted infix list. */
-          Hd_val (q) = (header_t) ((word) p | 2);
-          /* Change block header's tag to Infix_tag, and change its size
-             to point to the infix list. */
-          *hp = Make_ehd (Wosize_bhsize (q - val), Infix_tag, 3, (uintnat) 0);
-        }else{
-          CAMLassert (Tag_ehd (*hp) == Infix_tag);
-          /* Point the last of this infix list to the current first infix
-             list of the block. */
-          *p = (word) &Field (val, Wosize_ehd (*hp)) | 1;
-          /* Point the head of this infix list to the above. */
-          Hd_val (q) = (header_t) ((word) p | 2);
-          /* Change block header's size to point to this infix list. */
-          *hp = Make_ehd (Wosize_bhsize (q - val), Infix_tag, 3, (uintnat) 0);
-        }
+  header_t h;
+
+  CAMLassert (((uintnat) p & 3) == 0);
+
+  if (Is_block (q)){
```

Comment:
Beware, this piece of code should work even in naked-pointers mode, where we do not know at this point that `q` is a heap pointer. So I think that you need to check for `Is_in_value_area(q)` (or `Is_in_heap(q)`) here.

```suggestion
  if (Is_block (q) && Is_in_value_area(q)){
```

Reply:
I agree with @jhjourdan that this code should work in naked-pointers mode too.  Whether his suggested change is enough or not, I don't know.

Reply:
I do think this will be enough, it's the only place where we look at arbitrary fields and roots to find the pointers into the heap. Note that it's also the only place where we test for black headers.